warehouse/validations/validate.go

```diff
@@ -17,378 +22,549 @@ import (
+func (*dummyUploader) GetLocalSchema() warehouseutils.SchemaT           { return warehouseutils.SchemaT{} }
+func (*dummyUploader) UpdateLocalSchema(_ warehouseutils.SchemaT) error { return nil }
+func (*dummyUploader) ShouldOnDedupUseNewRecord() bool                  { return false }
+func (*dummyUploader) GetFirstLastEvent() (time.Time, time.Time)        { return time.Time{}, time.Time{} }
+func (*dummyUploader) GetLoadFileGenStartTIme() time.Time               { return time.Time{} }
+func (*dummyUploader) GetSampleLoadFileLocation(string) (string, error) { return "", nil }
+func (*dummyUploader) GetTableSchemaInWarehouse(string) warehouseutils.TableSchemaT {
+	return warehouseutils.TableSchemaT{}
+}
+
+func (*dummyUploader) GetTableSchemaInUpload(string) warehouseutils.TableSchemaT {
+	return warehouseutils.TableSchemaT{}
+}
+
+func (*dummyUploader) GetLoadFilesMetadata(warehouseutils.GetLoadFilesOptionsT) []warehouseutils.LoadFileT {
+	return []warehouseutils.LoadFileT{}
+}
+
+func (*dummyUploader) GetSingleLoadFile(string) (warehouseutils.LoadFileT, error) {
+	return warehouseutils.LoadFileT{}, nil
+}
+
+func (m *dummyUploader) GetLoadFileType() string {
+	return warehouseutils.GetLoadFileType(m.dest.DestinationDefinition.Name)
+}
+
+func (m *dummyUploader) UseRudderStorage() bool {
+	return misc.IsConfiguredToUseRudderObjectStorage(m.dest.Config)
+}
+
+func NewDestinationValidator() DestinationValidator {
+	return &destinationValidationImpl{}
+}
+
+func (v *destinationValidationImpl) Validate(dest *backendconfig.DestinationT) *model.DestinationValidationResponse {
+	return validateDestination(dest, "")
+}
+
+func validateDestinationFunc(dest *backendconfig.DestinationT, stepToValidate string) (json.RawMessage, error) {
+	return json.Marshal(validateDestination(dest, stepToValidate))
+}
+
+func validateDestination(dest *backendconfig.DestinationT, stepToValidate string) *model.DestinationValidationResponse {
 	var (
-		dest     backendconfig.DestinationT
-		destID   string
-		destType string
+		destID          = dest.ID
+		destType        = dest.DestinationDefinition.Name
+		stepsToValidate []*model.Step
+		validator       Validator
+		err             error
 	)
 
-	dest = ct.infoRequest.Destination
-	destID, destType = dest.ID, dest.DestinationDefinition.Name
-
-	pkgLogger.Infof("Validating destination configuration for destinationId: %s, destinationType: %s, step: %s",
-		destID,
-		destType,
-		step,
+	pkgLogger.Infow("validate destination configuration",
+		logfield.DestinationID, destID,
+		logfield.DestinationType, destType,
+		logfield.DestinationRevisionID, dest.RevisionID,
+		logfield.WorkspaceID, dest.WorkspaceID,
+		logfield.DestinationValidationsStep, stepToValidate,
 	)
 
-	resp := DestinationValidationResponse{}
 	// check if req has specified a step in query params
-	if step != "" {
-		stepI, err := strconv.Atoi(step)
+	if stepToValidate != "" {
+		stepI, err := strconv.Atoi(stepToValidate)
```

Comment:
Why don't we accept this as an integer? We can get away with all this type cast right

Reply:
This comes as [query params](https://github.com/rudderlabs/cp-router/blob/8d929099e5cae0f707fdc0ec87d623675898003c/api/warehouse.go#L298), from config backend -> cp-router -> warehouse. Either we need to do this conversion at the cp-router or at the warehouse. Since the warehouse acts as a proxy, it's better we handle it at the warehouse.